Add draft capacity planning doc (#4008)
jupyterhub/jupyterhub: docs/source/admin/capacity-planning.md

# Capacity planning

> General capacity planning advice for JupyterHub is hard to give,

Author: Point taken! I generally follow "semantic line breaks" for exactly that reason, but break on clauses rather than only full sentences. I can do sentences, though!

> General capacity planning advice for JupyterHub is hard to give,

Reviewer: I am also a fan of semantic breaks :-) I've just found that for some reason, it is more common for me to want to edit at the sentence-level rather than the clause level, so diffing one line corresponding to a sentence is the compromise I use. Either one is fine, I just wanted to note it

> General capacity planning advice for JupyterHub is hard to give,

Author: Yeah, I mostly thought about reading/editing a fixed diff, but sentences makes total sense in the context of tools like GitHub suggestions. I hadn't thought of that.

because it depends almost entirely on what your users are doing,
and what JupyterHub users do varies _wildly_ in terms of resource consumption.

**There is no single answer to "I have X users, what resources do I need?" or "How many users can I support with this machine?"**

Here are three _typical_ Jupyter use patterns that require vastly different resources:

- negligible resources because computation is mostly idle,
  e.g. students learning programming for the first time
- very intense, sustained load, e.g. training machine learning models
- _mostly_ idle, but needs a lot of resources for short periods of time

>   (interactive research often looks like this)

Author: I like this, though I'm not sure the 'scaling out/up'  are the right labels for the first two. These are descriptions of _individual user_ behavior, not cluster behavior. Maybe "Reading/learning" and "Production code"?

>   (interactive research often looks like this)

Reviewer: good point - in my mind, the amount of resources needed is _usually_ correlated with the total number of users, but I agree it's not strictly true

>   (interactive research often looks like this)

Author: > amount of resources needed is usually correlated with the total number of users

Yes, though it's more directly related to concurrent users. I just added a section on concurrent users which discusses some of this and links the idle culler.

I think there are 3 questions I expect most people coming this doc to have:

- I have X resources, how many users can I handle?
- I have Y users, what total resources do I need?
- What do my individual users need (already looking for what to put in the z2jh singleuser resource fields)?

and I want to get right away to the answer to all three: "It depends. What are your users are doing?"

The framing of this whole doc should be: understanding what your users actually do is key to accurately allocating resources.

>   (interactive research often looks like this)

Author: fwiw, I went with:

- **Learning**
 - **Production code**
 - **Bursting**

On the other hand, if you are using ephemeral resources,
such as node pools in Kubernetes,
changing resource types costs close to nothing
because nodes can automatically be added or removed as needed.

Take that cost into account when you are picking how much memory or cpu to allocate to users.

Static resources (like [the-littlest-jupyterhub][]) provide for more **stable, predictable costs**,
but elastic resources (like [zero-to-jupyterhub][]) tend to provide **lower overall costs**
(especially when deployed with monitoring allowing cost optimizations over time),
but which are **less predictable**.

[the-littlest-jupyterhub]: https://the-littlest-jupyterhub.readthedocs.io

[zero-to-jupyterhub]: https://zero-to-jupyterhub.readthedocs.io

(limits-requests)=

### Limit vs Request

Many scheduling tools like Kubernetes have two separate ways of allocating resources to users.
A **Request** or **Reservation** describes how much resources are _set aside_ for each user.
Often, this doesn't have any practical effect other than deciding when a given machine is considered 'full'.
If you are using expandable resources like an autoscaling Kubernetes cluster,

> 'requesting' more resources than fit on currently running nodes is when a new node is launched and added to the pool (a cluster **scale-up event**).

Author: I found it a little hard to write ;)

If you are running on a single VM, this describes how many users you can run at the same time, full stop.

A **Limit**, on the other hand, actually enforces a limit to how much resources any given user can consume.
We'll see more information on what happens when users try to exceed their limits [below](oversubscription).

In the strictest, safest case, you can have these two numbers be the same.
That means that each user is _limited_ to fit within the resources allocated to it.
This avoids **[oversubscription](oversubscription)** of resources (allowing use of more than you have available),
at the expense (in a literal, this-costs-money sense) of reserving lots of usually-idle capacity.

But when deploying JupyterHub,
you will likely find that a relatively small fraction of users use lots more resources than others,
making oversubscription attractive (to a point).

Having a gap between the request and the limit means you can fit a number of _typical_ users on a node (based on the request),
but still limit how much a runaway user can gobble up for themselves.

(oversubscription)=

### Oversubscribed CPU is okay, running out of memory is bad

> An important consideration when assigning resources to users is:
> 
> > What happens when users need more than I've given them?
> 
> A good summary to keep in mind: **when tasks don't get enough CPU, things are slow.
> When they don't get enough memory, things are broken.**
> Which means it's very important that users have enough memory,
> but much less important that they always have exclusive access to all the CPU they can use.

Reviewer: ```suggestion
An important consideration when assigning resources to users is: **What happens when users need more than I've given them?**

A good summary to keep in mind:

> When tasks don't get enough CPU, things are slow.
  When they don't get enough memory, things are broken.

This means it's **very important that users have enough memory**,
but much less important that they always have exclusive access to all the CPU they can use.
```

This highlights the "important part" a bit more clearly

Author: Yeah, I was surprised by the lack of block quote formatting. I think I have a bug report for our sphinx theme on the way.

This relates to [Limits and Requests](limits-requests),
because these are the consequences of your limits and/or requests not matching what users actually try to use.

A table of mismatched resource allocation situations and their consequences:

> | issue                                                    | consequence                                                                           |

Reviewer: this is great

| Requests too high                                        | Unnecessarily high cost and/or low capacity.                                          |
| CPU limit too low                                        | Poor performance experienced by users                                                 |
| CPU oversubscribed (too-low request + too-high limit)    | Poor performance across the system; may crash, if severe                              |
| Memory limit too low                                     | Servers killed by Out-of-Memory Killer (OOM); lost work for users                     |
| Memory oversubscribed (too-low request + too-high limit) | System memory exhaustion - all kinds of hangs and crashes and weird errors. Very bad. |

Note that the 'oversubscribed' problem case is where the request is lower than _typical_ usage,
meaning that the total reserved resources isn't enough for the total _actual_ consumption.
This doesn't mean that _all_ your users exceed the request,
just that the _limit_ gives enough room for the _average_ user to exceed the request.

> ### Example case for oversubscribe memory

Reviewer: Maybe we should point out somewhere really clearly that the most important number is the **Average CPU / Memory used per user on a node**. That's ultimately the number that decides whether you hit your node resource limits

Author: ```suggestion
### Example case for oversubscribing memory
```

Take for example, this system and sampling of user behavior:

- System memory = 8G
- memory request = 1G, limit = 3G
- typical 'heavy' user: 2G
- typical 'light' user: 0.5G

This will assign 8 users to those 8G of RAM (remember: only requests are used for deciding when a machine is 'full').
As long as the total of 8 users _actual_ usage is under 8G, everything is fine.
But the _limit_ allows a total of 24G to be used,
which would be a mess if everyone used their full limit.
But _not_ everyone uses the full limit, which is the point!

This pattern is fine if 1/8 of your users are 'heavy' because _typical_ usage will be ~0.7G,
and your total usage will be ~5G (1 × 2 + 7 × 0.5 = 5.5).

> But if _50%_ of your users are 'heavy' you have a problem because that means your users will be trying to use 10G (4 × 2 + 4 × 0.5 = 10),

Reviewer: ```suggestion
But if _50%_ of your users are 'heavy' you have a problem because that means your users will be trying to use 10G (`4 × 2 + 4 × 0.5 = 10`),
```

I find it cleaner to read arithmetic in docs as inline code blocks, wdyt? (I do the same for numbers that are used often as well)

which you don't have.

> You can make guesses at these numbers, but the only _real_ way to get them is to measure (more [below](measuring)).

Reviewer: ```suggestion
You can make guesses at these numbers, but the only _real_ way to get them is to measure (see [](measuring)).
```

I think it's good to use "content-ful" link titles, rather than vague ones like `below` - by removing the title entirely it'll auto-populate the title with the section heading

but often does not significantly affect the _memory_.

Ways to think about this:

- More idle users means unused CPU.
  This generally means setting your CPU _limit_ higher than your CPU _request_.
- What do your users do when they _are_ running code?
  Is it typically single-threaded local computation in a notebook?
  If so, there's little reason to set a limit higher than 1 CPU core.
- Do typical computations take a long time, or just a few seconds?
  Longer typical computations means it's more likely for users to be trying to use the CPU at the same moment,
  suggesting a higher _request_.
- Even with idle users, parallel computation adds up quickly - one user fully loading 4 cores and 3 using almost nothing still averages to more than a full CPU core per user.
- Long-running intense computations suggest higher requests.

Again, using mybinder.org as an example—we run around 100 users on 8-core nodes,
and still see fairly _low_ overall CPU usage on each user node.
The limit here is actually Kubernetes' pods per node, not memory _or_ CPU.
This is likely a extreme case, as many Binder users come from clicking links on webpages
without any actual intention of running code.

> ![mybinder.org node CPU usage is low with 50-150 users sharing just 8 cores](../images/mybinder-load5.png)
> 

Reviewer: ````suggestion
```[figure} ../images/mybinder-load5.png
mybinder.org node CPU usage is low with 50-150 users sharing just 8 cores
```

````

## More tips

### Start strict and generous, then measure

A good tip, in general, is to give your users as much resources as you can afford that you think they _might_ use.
Then, use resource usage metrics like prometheus to analyze what your users _actually_ need,
and tune accordingly.
Remember: **Limits affect your user experience and stability. Requests mostly affect your costs**.

For example, a sensible starting point (lacking any other information) might be:

```yaml
request:
  cpu: 0.5
  mem: 2G
limit:
  cpu: 1
  mem: 2G
```

(more memory if significant computations are likely - machine learning models, data analysis, etc.)

Some actions

- If you see out-of-memory killer events, increase the limit (or talk to your users!)
- If you see typical memory well below your limit, reduce the request (but not the limit)
- If _nobody_ uses that much memory, reduce your limit
- If CPU is your limiting scheduling factor and your CPUs are mostly idle,
  reduce the cpu request (maybe even to 0!).
- If CPU usage continues to be low, increase the limit to 2 or 4 to allow bursts of parallel execution.

(measuring)=

> ### Measuring user resource consumption

Reviewer: I'd also recommend adding a section on cloud provider cost monitoring and limits services. I think these are often the simplest way to limit your cloud bill (e.g. "Have AWS send me an email if I hit XXX spending trajectory on week 3 of the month")

Author: Good call. Added.
